openstates/al/bills.py

```diff
@@ -39,222 +43,357 @@ def _categorize_action(action):
     return 'other'
 
 class ALBillScraper(BillScraper):
-
     jurisdiction = 'al'
-
-    def refresh_session(self):
-        url = ('http://alisondb.legislature.state.al.us/acas/ACASLoginFire.asp'
-               '?SESSION=%s') % self.session_id
-        html = self.urlopen(url)
-
-    def scrape(self, chamber, session):
-        self.session_id = self.metadata['session_details'][session]['internal_id']
-        self.base_doc_url = 'http://alisondb.legislature.state.al.us/acas/searchableinstruments/%s/PrintFiles/' % session
-
-        chamber_piece = {'upper': 'Senate',
-                         'lower': 'House+of+Representatives'}[chamber]
-        # resolutions
-        res_url = ('http://alisondb.legislature.state.al.us/acas/SESSResosBySe'
-                   'lectedMatterTransResults.asp?WhichResos=%s&TransCodes='
-                   '{All}&LegDay={All}') % chamber_piece
-        self.scrape_for_bill_type(chamber, session, res_url)
-
-        bill_url = ('http://alisondb.legislature.state.al.us/acas/SESSBillsByS'
-                    'electedMatterTransResults.asp?TransCodes={All}'
-                    '&LegDay={All}&WhichBills=%s') % chamber_piece
-        self.scrape_for_bill_type(chamber, session, bill_url)
-
-
-    def scrape_for_bill_type(self, chamber, session, url):
-
-        self.refresh_session()
-
-        html = self.urlopen(url)
-        doc = lxml.html.fromstring(html)
-
-        # bills are all their own table with cellspacing=4 (skip first)
-        bill_tables = doc.xpath('//table[@cellspacing="4"]')
-        for bt in bill_tables[1:]:
-
-            # each table has 3 rows: detail row, description, blank
-            details, desc, _ = bt.xpath('tr')
-
-            # first <tr> has img, button, sponsor, topic, current house
-            #   current status, committee, committee2, last action
-            tds = details.xpath('td')
-            if len(tds) == 9:
-                # middle _, _ is chamber, last action
-                _, button, sponsor, subject, _, _, com1, com2, _ = tds
-            elif len(tds) == 8:
-                # middle _ is last action
-                _, button, sponsor, subject, _, com1, com2, _ = tds
+    CHAMBERS = {'H': 'lower', 'S': 'upper'}
+    DATE_FORMAT = '%m/%d/%Y'
+
+    # Due to a scrapelib bug, a requests Session is necessary to maintain cookies
+    # Retries are also required since the server periodically returns errors
+    # Once this is fixed, roll back to vanilla scrapelib GETs and POSTs
+    s = requests.Session()
+
+    def sget(self, **request_args):
+        self.log('Special GET request to {}'.format(request_args['url']))
+        retry_counter = 0
+        RETRIES = 5
+        while retry_counter < RETRIES:
+            retry_counter += 1
+            time.sleep(retry_counter ** 2)
```

Comment:
This is built into scrapelib

On Fri, Feb 27, 2015 at 10:47 AM, <notifications@github.com> wrote:

> In openstates/al/bills.py
> <https://github.com/sunlightlabs/openstates/pull/590#discussion_r25514823>
> :
>
> > -                _, button, sponsor, subject, _, com1, com2, _ = tds
> > +    CHAMBERS = {'H': 'lower', 'S': 'upper'}
> > +    DATE_FORMAT = '%m/%d/%Y'
> > +
> > +    # Due to a scrapelib bug, a requests Session is necessary to maintain cookies
> > +    # Retries are also required since the server periodically returns errors
> > +    # Once this is fixed, roll back to vanilla scrapelib GETs and POSTs
> > +    s = requests.Session()
> > +
> > +    def sget(self, **request_args):
> > +        self.log('Special GET request to {}'.format(request_args['url']))
> > +        retry_counter = 0
> > +        RETRIES = 5
> > +        while retry_counter < RETRIES:
> > +            retry_counter += 1
> > +            time.sleep(retry_counter ** 2)
>
> +        RETRIES = 5+        while retry_counter < RETRIES:+            retry_counter += 1+            time.sleep(retry_counter ** 2)
>
> Also we could nom this guy out, and use the Backoff Mixin
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sunlightlabs/openstates/pull/590/files#r25514823>.
>



-- 

James Turk
Labs Director | Sunlight Foundation <http://www.sunlightfoundation.com/>
202-558-8723 | @JamesTurk <https://twitter.com/jamesturk>

[image: Sunlight Foundation] <http://sunlightfoundation.com/>